src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSource.cs

```diff
@@ -593,20 +630,43 @@ private static SuggestedActionSetPriority GetSuggestedActionSetPriority(CodeActi
                 SnapshotSpan range,
                 CancellationToken cancellationToken)
             {
-                // Explicitly hold onto below fields in locals and use these locals throughout this code path to avoid crashes
-                // if these fields happen to be cleared by Dispose() below. This is required since this code path involves
-                // code that can run asynchronously from background thread.
-                var view = _textView;
-                var buffer = _subjectBuffer;
                 var provider = _owner;
 
-                if (view == null || buffer == null || provider == null)
+                if (IsDisposed)
                 {
+                    // We've already been disposed.  No point in continuing.
                     return false;
                 }
 
-                using (var asyncToken = provider.OperationListener.BeginAsyncOperation("HasSuggestedActionsAsync"))
+                using (var asyncToken = _owner.OperationListener.BeginAsyncOperation("HasSuggestedActionsAsync"))
                 {
+                    // Acquire the user's selection up front, before we do any async work, so that we can
+                    // directly grab it when we're on the UI thread.  That way we don't have any reentrancy
+                    // blocking concerns if VS wants to block on this call (for example, if the user 
+                    // explicitly invokes the 'show smart tag' command).
+                    //
+                    // This work must happen on the UI thread as it needs to access the _textView's mutable
+                    // state.
+                    TextSpan? selection = null;
+                    if (IsForeground())
+                    {
+                        selection = TryGetCodeRefactoringSelection(range);
+                    }
+                    else
```

Comment:
These calls still come in on the background normally (and are used to determine if a lightbulb should even be shown).  They come in on the foreground if the BG call hasn't completed and hte user explicit does a ctrl-dot.  In that case they want to get the data immediately so they can expand the list and show the right lightbulb icon.  

I will doc as such!